File: week01/Assignment01/Assignment01_Code02.py
# Assignment01_Code02.py

def function(val):
    stack = []
    for char in val:
        if char == '(': stack.append(1)
        else:
            if not stack: return "NO"
            stack.pop()
    
    return "YES" if not stack else "NO"

t = int(input())
for _ in range(t):
    val = input().strip()
    print(function(val))

File: week01/Assignment01/Assignment01_Code02.cpp
// Assignment01_Code02.cpp

#include <iostream>
#include <stack>
using namespace std;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	int t;
	cin >> t;
	while (t--) {
		string val;
		cin >> val;

		stack<int> s;
		bool isValid = true;
		for (char x : val) {
			if (x == '(') s.push(1);
			else {
				if (s.empty()) {
					isValid = false;
					break;
				}
				s.pop();
			}
		}

		cout << (isValid && s.empty() ? "YES" : "NO") << '\n';
	}

	return 0;
}